### PR TITLE
RegexBuilder Options Port: Enable Overrides for RaBuilder

### PR DESCRIPTION
**Pull request by jonperry-dev**
# Case Insensitive Builder Feature Port

This is a small add-in to allow for options to be passed in as boolean values and mutate the `Config` through the `RegexBuilder`. This is primarily to align with the standard Regex crate for their own [RegexBuilder](https://docs.rs/regex/latest/src/regex/builders.rs.html#322-325).

# Changes

- Added a field to the `RegexOptions` to keep a state of `regex_automata::utils::syntax::Config` internally to disseminate the proper flag overrides
- Added public method to `Regexbuilder` called `case_insensitive` 
		- This will allow for forcing of a flag if not present in the raw str
- Added a call to the `RaRegexBuilder` which will consume this option using `.syntax( ... )`
- Added tests into `lib.rs` to make sure the override works as intended 

### Examples

Currently, we can write in flags into the raw string like this:
```rust 
use fancy_regex::RegexBuilder;

let re = r"(?i)FOO BAR";
let regex = RegexBuilder::new(re).build()?;
assert!(regex.is_match("foo bar").unwrap_or_default());
```
This change is to enable builder functionality like this:
```rust 
use fancy_regex::RegexBuilder;

let re = r"FOO BAR";
let regex = RegexBuilder::new(re)
	.case_insensitive(true)
	.build()?;
assert!(regex.is_match("foo bar").unwrap_or_default());
```

# Notes

Currently, `fancy-regex` doesn't support flags such as `(?-i)` and that is likely something we should enable in the future. The correlation of that type of flag to this feature add ties to the builder `case_insensitive` setter method. This method should override any raw input data.

### CODE REVIEW

File: src/compile.rs
Comment (robinst) [inline src/compile.rs:502]: Is this the only place where we actually construct a case insensitive matcher at the moment?

Comment (jonperry-dev) [reply src/compile.rs:502]: To the best of my knowledge and search, I believe this is the only place that we need this.

Comment (jonperry-dev) [reply src/compile.rs:502]: @robinst Just wanted to bump this, was there anything you feel like is missing currently for `lowercase` specific building? 

I'd like to iterate over the rest of the builder functionality as a future works.

Comment (robinst) [reply src/compile.rs:502]: Hey @jonperry-dev, thanks for your patience. Yeah, let's go ahead with this. Can you rebase the PR? The builds seem to have expired, sorry about that.

**Review by robinst — APPROVED** [review]

File: src/compile.rs
Comment (jonperry-dev) [reply src/compile.rs:502]: No worries at all! Thanks for getting back to me. 

Just updated, should be good to go.

I hope everything is well! 🙂

Comment (jonperry-dev) [reply src/compile.rs:502]: @robinst I think its now waiting for maintainer approval on CI. I did a few more changes: 
- Cleaned up the code
- had a `cargo fmt` change
- moved tests to its own file (plan on making more extensions to the options)

**Comment by jonperry-dev** [comment]
@robinst I'm not sure if that coverage failure with line 505 is an issue? I'm not entirely sure what its looking for there.

**Comment by robinst** [comment]
> @robinst I'm not sure if that coverage failure with line 505 is an issue? I'm not entirely sure what its looking for there.

Seems like tarpaulin (the tool we use for coverage) doesn't analyze that correctly. (Maybe bumping the version would help, not sure.)

**Comment by robinst** [comment]
Thanks, merged!

**Comment by robinst** [comment]
Released in 0.14.0: https://github.com/fancy-regex/fancy-regex/blob/main/CHANGELOG.md#0140---2024-10-24

**Pull request merged**
